src/main/scala/cognite/spark/v1/FileContentRelation.scala

```diff
@@ -0,0 +1,141 @@
+    override val sqlContext: SQLContext)
+    extends BaseRelation
+    with TableScan
+    with PrunedFilteredScan
+    with Serializable
+    with WithSizeLimit {
+
+  override val sizeLimit: Long = 5 * FileUtils.ONE_GB
+
+  @transient lazy val client: GenericClient[IO] =
+    CdpConnector.clientFromConfig(config)
+
+  @transient private lazy val sttpFileContentStreamingBackendResource
+    : Resource[IO, SttpBackend[IO, Fs2Streams[IO] with WebSockets]] =
+    for {
+      dispatcher <- Dispatcher.parallel[IO]
+      backend <- Resource.make(
+        IO(
+          AsyncHttpClientFs2Backend
+            .usingClient[IO](SttpClientBackendFactory.create("file content download"), dispatcher))
+      )(backend => backend.close())
+    } yield backend
+
+  private lazy val dataFrame: DataFrame = createDataFrame(sparkSession = sqlContext.sparkSession)
+
+  override def schema: StructType =
+    dataFrame.schema
+
+  def createDataFrame(sparkSession: SparkSession): DataFrame = {
+    val rdd: RDD[String] = new RDD[String](sparkSession.sparkContext, Nil) with Serializable {
+
+      import cognite.spark.v1.CdpConnector.ioRuntime
+
+      val maxParallelism = 1
+      override def compute(split: Partition, context: TaskContext): Iterator[String] = {
+
+        val validUrl = for {
+          mimeType <- client.files.retrieveByExternalId(fileExternalId).map(_.mimeType)
+          downloadLink <- client.files
+            .downloadLink(FileDownloadExternalId(fileExternalId))
+            .map(_.downloadUrl)
+          _ <- IO.pure(
+            if (!downloadLink.startsWith("https")) {
+              throw new CdfSparkException("File storage is not using https protocol")
+            }
+          )
+          isFileWithinLimits <- isFileWithinLimits(downloadLink)
+        } yield {
+          if (mimeType.isDefined && !Seq(
+              "application/jsonlines",
+              "application/x-ndjson",
+              "application/jsonl").contains(mimeType.get))
+            throw new CdfSparkException("Wrong mimetype. Expects application/jsonlines")
+          if (!isFileWithinLimits)
+            throw new CdfSparkException(
+              f"File size above size limit, or file size header absent from head request")
+          downloadLink
+        }
+
+        StreamIterator(
+          Stream
+            .eval(validUrl)
+            .flatMap(readUrlContent),
+          maxParallelism,
+          None
+        )
+
+      }
+
+      override protected def getPartitions: Array[Partition] =
+        Array(CdfPartition(0))
+    }
+    import sparkSession.implicits._
+    val dataset = rdd.toDS()
+
+    if (inferSchema) {
+      sqlContext.sparkSession.read.json(dataset)
+    } else {
+      dataset.toDF()
+    }
+  }
+
+  private def readUrlContent(link: String): Stream[IO, String] =
+    Stream.resource(sttpFileContentStreamingBackendResource).flatMap { backend =>
+      val request = basicRequest
```

Comment:
maybe a better idea is to start reading and use a trick like suggested here: https://softwaremill.community/t/limit-the-download-size-of-an-sttp-client/270 to limit the size?

Since content length does not work with gzip (see here: https://stackoverflow.com/questions/3819280/content-length-when-using-http-compression) that might be a better approach, or we could use both (if content-length > 5gb, don(t bother downloading, if it's < download but still enforce limit using the hack in the first link?).

Reply:
worth a try is also doing HEAD request without gzip support, if server would send uncompressed length that way we're good

but otherwise maybe when we don't have info on file size indeed we could do limit how much we read from it (maybe on line stream level as ideally we should not damage the last line, or always drop last line for limit exceeded case)

Reply:
We can skip the headers, but it's unpractical to actually test if it's implemented correctly since there is a bunch of different services depending on the cluster...

Reply:
The vendors list isn't that long, it would be good info on which ones support compression and which ones report content-length (and maybe depending on compression on/off). If we have the "size unknown" fallback that tries and fails if too big might be a good path, but even then it might be good to know how often it would fail first doing the read

Reply:
For S3: Sounds like it *should* work
For Gcloud, I stumbled upon this: https://cloud.google.com/storage/docs/transcoding#decompressive_transcoding

Which means it's possible to use decompressive transcoding to download a file that is actually stored as a gzip (this can happen when the content-encoding is set to gzip when uploading)

But when decompressive transcoding is used, the head request will not return a content length`

`When an object meets these two criteria, it undergoes decompressive transcoding when served, and the response containing the object does not contain a Content-Encoding or Content-Length header.`

..Back to the drawing board I guess :/

Reply:
we also can start with runtime limiting and see the best way to handle compression&contentLength before making it GA